### DISCUSSION TITLE
how to start write some code?

### DISCUSSION THREAD

**Discussion by tu1620808124** [libMesh/libmesh#3451] [General]
I just started learnning libmesh, I have installed, and use the ```make check``` learned five basic example, but there no C code in the directory```/build/examples/instroduction/instroduction_ex1```.  I want to  modify some parameters (such as geometry  modification, etc.), but how I can do to compile and run?

**Comment by jwpeterson** [comment]
Hi,

You are in the "build" directory, while the source code is in the top-level `examples/introduction/introduction_ex1` directory. If you make modifications to the example source code and then run e.g. `make example-opt` from the corresponding build directory for that example, it will pick up your new changes.

**Comment by tu1620808124** [comment]
Hi, 
As you said, it is true, but I am still confused, usually the C code we wrote can be compiled and run. But here seems to be a lot of files in two different directories (for example, example 2, there is a run.sh file), it seems that they are connected through Makefile, can briefly talk about the how these files works?

**Comment by jwpeterson** [comment]
The run.sh file (which lives in the source tree, because it is maintained by humans) contains command line arguments that are appropriate for each example. If you do `make run` from the build directory for a specific example, you will see that it runs all the commands from the run.sh file.

> can briefly talk about the how these files works?

The setup used by the libmesh example codes is typical for projects that use the `autotools` (autoconf, automake, etc.) and it splits up compiler-generated files (object files, executables, etc.) into a "build" directory which is separate from the "source" directory. I would not recommend investing too much time into figuring out how it really works, since your own project will likely not use autotools (at least initially). 

Assuming you have all your code in a single file called `my_code.C`, you can write a simple Makefile to compile it using the `libmesh-config` script, for example:
```
include $(LIBMESH_DIR)/Make.common

all: my_code.C
        `"$(LIBMESH_DIR)/bin/libmesh-config" --cxx` -o my_code my_code.C `"$(LIBMESH_DIR)/bin/libmesh-config" --cxxflags --include --ldflags --libs`
```
where `$(LIBMESH_DIR)` is the prefix you set during configure.

**Reply by tu1620808124** [reply]
ok, thank you so much!
